### PR TITLE
Modular extensions

### LINKED ISSUE #428: Make all build dependencies optional
This feature is related to the following recent issues with other institutions trying to build TomoPy with the new build system and dependencies:

- https://github.com/NSLS-II/docs/pull/112
- https://github.com/conda-forge/tomopy-feedstock/pull/21
- I also had a request for a no-MKL build (for licensing reasons), but I can't find the Issue.

It seems that people are both not expecting the hard OpenCV and MKL build dependencies, but might not need or want them.

**Describe the solution you would like.**
We should make all of the build dependencies, MKL, OpenCV, and CUDA optional, and return a helpful error if users try to use the algorithms that depend on these libraries which explains that they need to have these libraries installed at build-time.

**Describe alternatives you've considered.**
The alternative is to do nothing.

**Additional context.**
CUDA is already optional at build time, but as described in #422, the CUDA methods will fall back instead of returning an error if CUDA fails.

### LINKED ISSUE #448: Minimal dependencies to build tomopy for reconstruction
**Describe the problem**
We are looking at the best way of adding TomoPy to Tomviz.  We have a Python environment with Python 3.7, SciPy, NumPy, and have a PR to add h5py. The dependency set looked larger than expected, including OpenCV and several other components. What is the minimal set of dependencies/build configuration to support reconstruction?

**Expected behavior (if applicable)**
Ideally no big new dependencies, perhaps we need to understand how to turn off optional features.

**Helpful images (if applicable)**
Add screenshots to help explain your problem.

**Platform Information:**
 - OS: We want o support Linux, macOS, and Windows
 - Python Version 3.7
 - TomoPy Version Latest/master

**Additional context**
This is really us exploring  packaging, another approach is to use a Docker container with a miniconda starting point.

### LINKED ISSUE #452: 2020.04.26 Intern - SULI
I am [writing a project proposal](https://docs.google.com/document/d/1fYKdYe_87yAkXHk-QgIB0YxkD3v6H1d4V29qO-LakZc/edit?usp=sharing) for an undergraduate intern to refactor TomoPy. My current proposal addresses the following issues: #428, #449, #448, #422

Since, this is a very public project, I am posting the proposal for public feedback, and so no one is surprised in a few months when code is moved around. Basically, if you have an idle branch that you are planning to merge, please open a pull-request before **March 1, 2020**. This project may include API breaking changes in the Python layer, and it will definitely have API breaking changes for the C layer.

If anyone knows any undergrads who may be particularly suited to this project, please contact me via email. My contact information is available on the Argonne National Laboratory public website. I'm not going to link it directly because I don't want spam.

### PR DESCRIPTION
**Pull request by dreycenfoiles**
Previously, TomoPy has had the required dependencies OpenCV and MKL. These libraries provide CPU acceleration and the gridrec reconstruction algorithm, respectively. However, users of TomoPy may not have these dependencies installed and may not even need the functionality provided by them, so it does not make sense for them to be required during the TomoPy build process.

The best way to make the dependencies optional was to modularize TomoPy.  This is accomplished by breaking up the shared-object library file that was originally created into five distinct library files that have separate dependencies from each other. These libraries are "accel" (the accelerated MLEM and SIRT functions), "gridrec" (the gridrec reconstruction algorithm), as well as "misc", "prep" and "recon", which contain the C implementations of functions required in the Python files under the same directory titles. Under this new file organization, if MKL or OpenCV are not found, the corresponding shared-library object file will not be able to build, but the other library files still will. 

This PR also has organized the header files so that they are now divided into one of five folders depending on which library they provide forward declarations for.

It has also turned `extern.py` into a module with each of the five library files being one of its submodules where the corresponding library file is opened by Python. 

This removed functions from the headers that are not meant to be in the public API, since it was not possible to remove all non-public functions from the external headers, internal headers were created to provide forward declarations when multiple files rely on a function. 

Closes #428, closes  #448. closes #452

### CODE REVIEW

File: cmake/Modules/BuildSettings.cmake [hunk 1]
```diff
@@ -18,7 +18,7 @@ set(CMAKE_CXX_COMPILE_FEATURES
 #
 set(CMAKE_INSTALL_MESSAGE LAZY)
 set(CMAKE_C_STANDARD 11 CACHE STRING "C language standard")
-set(CMAKE_CXX_STANDARD 11 CACHE STRING "CXX language standard")
+set(CMAKE_CXX_STANDARD 14 CACHE STRING "CXX language standard")
```
Comment (carterbox) [inline cmake/Modules/BuildSettings.cmake:21]: Why is this change necessary?

File: cmake/Modules/Packages.cmake [hunk 2]
```diff
@@ -40,31 +40,39 @@ if(PYTHON_EXECUTABLE)
         ${PYTHON_ROOT_DIR}/include)
 endif()
 
-
 ################################################################################
 #
-#        MKL (required)
+#        OpenCV (required for CPU acceleration)
 #
 ################################################################################
```
Comment (carterbox) [inline cmake/Modules/Packages.cmake:47]: This OpenCV and MKL section were swapped in your branch. Reswap them so that the git diff is cleaner (to avoid unnecessary changes upstream).

File: source/libtomo/gridrec/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    PREFIX ""
+    RUNTIME_OUTPUT_DIRECTORY    ${OUTPUT_PATH}/tomopy/util/extern
+    LIBRARY_OUTPUT_DIRECTORY    ${OUTPUT_PATH}/tomopy/util/extern
+    ARCHIVE_OUTPUT_DIRECTORY    ${OUTPUT_PATH}/tomopy/util/extern
+    INCLUDE_DIRECTORIES         "${TARGET_INCLUDE_DIRECTORIES}"
+    C_STANDARD                  ${CMAKE_C_STANDARD}
+    C_STANDARD_REQUIRED         ${CMAKE_C_STANDARD_REQUIRED}
+    CXX_STANDARD                ${CMAKE_CXX_STANDARD}
+    CXX_STANDARD_REQUIRED       ${CMAKE_CXX_STANDARD_REQUIRED}
+    ${CUDA_PROPERTIES})
```
Comment (carterbox) [inline source/libtomo/gridrec/CMakeLists.txt:47]: For these new CMakeLists, let's clean them up so they aren't creating unnecessary links. For example, gridrec does not need `${CUDA_PROPERTIES}` and most of the libraries do not need to set a `CXX_STANDARD`.

File: source/libtomo/misc/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+# core sources
+set(MISC_LIBRARY_SOURCES ${misc_tomo_headers} ${misc_tomo_sources})
+
+# create library
+add_library(libtomopy-misc SHARED ${MISC_LIBRARY_SOURCES})
+
+# link library
+target_link_libraries(libtomopy-misc PUBLIC ${EXTERNAL_LIBRARIES}
+    PRIVATE ${EXTERNAL_PRIVATE_LIBRARIES})
```
Comment (carterbox) [inline source/libtomo/misc/CMakeLists.txt:20]: Libraries like misc do not have any external dependencies, so they should not be linked against `${EXTERNAL_LIBRARIES}`.

File: cmake/Modules/BuildSettings.cmake (hunk 1 above)
Comment (dreycenfoiles) [reply cmake/Modules/BuildSettings.cmake:21]: We were running into segfaults and you found that it was fixed by setting the standard to 14. This was what you told me:

"I was playing with your build, and I observed that both cxx and c gridrec work for Linux if the CXX_STANDARD is set correctly (it must be at least 14 to enable std::complex). The accelerated functions cause a segmentation fault."

Comment (carterbox) [reply cmake/Modules/BuildSettings.cmake:21]: Hmmm... I don't trust past me. Let's recheck this.

**Comment by dreycenfoiles** [comment]
> Libraries like misc do not have any external dependencies, so they should not be linked against `${EXTERNAL_LIBRARIES}`. 

When I remove that line and compile, I receive errors about undefined functions like `downsample`, `upsample`, `pow`, and `sincos`.

File: source/libtomo/misc/CMakeLists.txt (hunk 4 above)
Comment (dreycenfoiles) [reply source/libtomo/misc/CMakeLists.txt:20]: When I remove the call to `${EXTERNAL_LIBRARIES}` and build, I get an error saying I have undefined functions liked `downsample`, `upsample`, `atan2`, and `sincos`.

File: cmake/Modules/BuildSettings.cmake (hunk 1 above)
Comment (dreycenfoiles) [reply cmake/Modules/BuildSettings.cmake]: I think past you was right. When building on Windows, I run into errors when the CXX_STANDARD is 11 when building gridrec. Setting the standard to 14 fixes the issue.

File: source/libtomo/gridrec/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    CXX_STANDARD                ${CMAKE_CXX_STANDARD}
+    CXX_STANDARD_REQUIRED       ${CMAKE_CXX_STANDARD_REQUIRED})
+
+target_compile_definitions(libtomopy-gridrec PUBLIC
+    ${${PROJECT_NAME}_DEFINITIONS})
+
+target_compile_options(libtomopy-gridrec PUBLIC
+    $<$<COMPILE_LANGUAGE:C>:${${PROJECT_NAME}_C_FLAGS}>
+    $<$<COMPILE_LANGUAGE:CXX>:${${PROJECT_NAME}_CXX_FLAGS}>
+    $<$<COMPILE_LANGUAGE:CUDA>:${${PROJECT_NAME}_CUDA_FLAGS}>)
```
Comment (carterbox) [inline source/libtomo/gridrec/CMakeLists.txt:54]:
```suggestion
    $<$<COMPILE_LANGUAGE:CXX>:${${PROJECT_NAME}_CXX_FLAGS}>)
```

File: source/libtomo/misc/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    C_STANDARD                  ${CMAKE_C_STANDARD}
+    C_STANDARD_REQUIRED         ${CMAKE_C_STANDARD_REQUIRED})
+
+target_compile_definitions(libtomopy-misc PUBLIC
+    ${${PROJECT_NAME}_DEFINITIONS})
+
+target_compile_options(libtomopy-misc PUBLIC
+    $<$<COMPILE_LANGUAGE:C>:${${PROJECT_NAME}_C_FLAGS}>
+    $<$<COMPILE_LANGUAGE:CXX>:${${PROJECT_NAME}_CXX_FLAGS}>
+    $<$<COMPILE_LANGUAGE:CUDA>:${${PROJECT_NAME}_CUDA_FLAGS}>)
```
Comment (carterbox) [inline source/libtomo/misc/CMakeLists.txt:45]:
```suggestion
target_compile_options(libtomopy-misc PUBLIC
    $<$<COMPILE_LANGUAGE:C>:${${PROJECT_NAME}_C_FLAGS}>)
```

File: source/libtomo/prep/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    C_STANDARD                  ${CMAKE_C_STANDARD}
+    C_STANDARD_REQUIRED         ${CMAKE_C_STANDARD_REQUIRED})
+
+target_compile_definitions(libtomopy-prep PUBLIC
+    ${${PROJECT_NAME}_DEFINITIONS})
+
+target_compile_options(libtomopy-prep PUBLIC
+    $<$<COMPILE_LANGUAGE:C>:${${PROJECT_NAME}_C_FLAGS}>
+    $<$<COMPILE_LANGUAGE:CXX>:${${PROJECT_NAME}_CXX_FLAGS}>
+    $<$<COMPILE_LANGUAGE:CUDA>:${${PROJECT_NAME}_CUDA_FLAGS}>)
```
Comment (carterbox) [inline source/libtomo/prep/CMakeLists.txt:33]:
```suggestion
target_compile_options(libtomopy-prep PUBLIC
    $<$<COMPILE_LANGUAGE:C>:${${PROJECT_NAME}_C_FLAGS}>)
```

File: source/libtomo/recon/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    target_link_libraries(libtomopy-recon PUBLIC m)
+endif()
+
+target_compile_definitions(libtomopy-recon PUBLIC
+    ${${PROJECT_NAME}_DEFINITIONS})
+
+target_compile_options(libtomopy-recon PUBLIC
+    $<$<COMPILE_LANGUAGE:C>:${${PROJECT_NAME}_C_FLAGS}>
+    $<$<COMPILE_LANGUAGE:CXX>:${${PROJECT_NAME}_CXX_FLAGS}>
+    $<$<COMPILE_LANGUAGE:CUDA>:${${PROJECT_NAME}_CUDA_FLAGS}>)
```
Comment (carterbox) [inline source/libtomo/recon/CMakeLists.txt:34]:
```suggestion
target_compile_options(libtomopy-recon PUBLIC
    $<$<COMPILE_LANGUAGE:C>:${${PROJECT_NAME}_C_FLAGS}>)
```

File: source/libtomo/gridrec/CMakeLists.txt (hunk 3 above)
Comment (carterbox) [reply source/libtomo/gridrec/CMakeLists.txt]: Double check this. I found some places in the suggestions below where CUDA properties were being set on libraries that did not use CUDA.
